### PR TITLE
Logistic regression online example

### PR DESCRIPTION
**Pull request by michael-smirnov**
# Logistic regression online example

Changes proposed in this pull request:
- created SYCL examples of logistic regression online algorithm using SGD solver and loss functions
- fixed initialization of starting point in logistic regression batch algorithm
- relaxed conditions of some asserts in logreg - they are valid now in the case of both batch and online computations

### CODE REVIEW

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    SGDMinibatchSolverPtr _solver;
+
+    LogLossPtr _logLoss;
+    CrossEntropyLossPtr _crossEntropyLoss;
+
+    daal_dm::NumericTablePtr _workPoint;
+    daal_dm::NumericTablePtr _learningRateTable;
+    daal_dm::NumericTablePtr _finalBetas;
+
+    size_t _batchRowCount;
```
Comment (PetrovKP) [inline examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp:213]: Can some fields be obtained from tables `_workPoint`, `_learningRateTable`, `_finalBetas` methods getNumberOfRows, getNumberOfColumns?

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    using LogLossPtr = daal::services::SharedPtr<LogLoss>;
+
+    using CrossEntropyLoss = daal_solver::cross_entropy_loss::Batch<Fptype>;
+    using CrossEntropyLossPtr = daal::services::SharedPtr<CrossEntropyLoss>;
+
+    using SGDMinibatchSolver = daal_sgd::Batch<Fptype, daal_sgd::miniBatch>;
+    using SGDMinibatchSolverPtr = daal::services::SharedPtr<SGDMinibatchSolver>;
+
+public:
+    LogisticRegressionOnline()
```
Comment (PetrovKP) [inline examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp:44]: I would suggest to align with batch

```suggestion
    LogisticRegressionOnline(size_t nClasses)
```

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        : _batchRowCount(0), _nFeatures(0), _nClasses(0), _nBetas(0), _nBetaCols(0), _isIntercept(false) {}
+
+    // Sets the algorithm parameters
+    void setParams(size_t nClasses,
+                   size_t nFeatures,
```
Comment (PetrovKP) [inline examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp:49]: This information can be obtained from the table when setting the input.

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+            }
+
+            _workPoint->releaseBlockOfRows(bd);
+        }
+    }
+
+private:
+    SGDMinibatchSolverPtr _solver;
+
+    LogLossPtr _logLoss;
```
Comment (PetrovKP) [inline examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp:206]: LogLossPtr  and CrossEntropyLossPtr  have a base class objective_function::Batch why not store one instance of this pointer?

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp [hunk 5]
```diff
@@ -0,0 +1,219 @@
+/* file: logistic_regression_online.hpp */
```
Comment (PetrovKP) [inline examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp:1]: in all examples we use the extension *.h for headers.

```suggestion
/* file: logistic_regression_online.h */
```

**Comment by PetrovKP** [comment]
You forgot to add new examples to the file: daal_lnx.lst, without this your examples will not run in CI.

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+public:
+    LogisticRegressionOnline()
+        : _batchRowCount(0), _nFeatures(0), _nClasses(0), _nBetas(0), _nBetaCols(0), _isIntercept(false) {}
+
+    // Sets the algorithm parameters
+    void setParams(size_t nClasses,
+                   size_t nFeatures,
+                   bool isIntercept,
+                   Fptype l1_penalty,
```
[the remark this answers is not in the document]
Comment (michael-smirnov) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp:51]: Do not see this in the documentation. The only thing I've found is that L1 is not supported on GPU: https://docs.oneapi.com/versions/latest/onedal/dev_guide/algorithms/logistic_regression/logistic-regression.html
Neither loss functions nor solver docs say that L1 regularization is not supported for SGD: https://docs.oneapi.com/versions/latest/onedal/dev_guide/algorithms/optimization-solvers/solvers/stochastic-gradient-descent-algorithm.html

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp (hunk 1 above)
Comment (michael-smirnov) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp:213]: No. `_workPoint` stores betas as 1D array, `_learningRateTable` has the shape of 1x1, `_finalBetas` are only allocated in finalize().

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp (hunk 2 above)
Comment (michael-smirnov) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp:44]: Do not see the reason for this. I would like to initialize all the parameters at once instead of separating them.

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp (hunk 3 above)
Comment (michael-smirnov) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp:49]: But I need it here to initialize table for betas.

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp (hunk 4 above)
Comment (michael-smirnov) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp:206]: Because as you see [here](https://github.com/oneapi-src/oneDAL/blob/e8e3085b6c69c595e1f29f13d852fe06e100bce7/examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp#L113) I must set input to the loss function. Base class do not allow to set it - `data` and `dependentVariables` are different enum values for logLoss and CrossEntropy. 
Moreover, in setParameter() I  set loss function parameters after extracting it from the solver. It is also cannot be done via base class. Rather than having the same `if`s with casts from base class to the original one I would like to make the code simpler.

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+    using CrossEntropyLoss = daal_solver::cross_entropy_loss::Batch<Fptype>;
+    using CrossEntropyLossPtr = daal::services::SharedPtr<CrossEntropyLoss>;
+
+    using SGDMinibatchSolver = daal_sgd::Batch<Fptype, daal_sgd::miniBatch>;
+    using SGDMinibatchSolverPtr = daal::services::SharedPtr<SGDMinibatchSolver>;
+
+public:
+    LogisticRegressionOnline()
+        : _batchRowCount(0), _nFeatures(0), _nClasses(0), _nBetas(0), _nBetaCols(0), _isIntercept(false) {}
```
[the remark this answers is not in the document]
Comment (michael-smirnov) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp]: Removed, thanks

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (michael-smirnov) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp]: I removed L1 regularization from parameters anyway

[the remark this answers is not in the document]
Comment (PetrovKP) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp]: I'm very surprised. We kind of added this information, @ShvetsKS didn't we?

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp (hunk 1 above)
Comment (PetrovKP) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp]: ok

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp (hunk 2 above)
Comment (PetrovKP) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp]: I see you divided on parts: constructors, set_params, set_inputs, compute, finalize by analogy with container of algorithms. So, in this case, it would look consistency. Otherwise, why not pass all these parameters directly to the constructor?

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp (hunk 4 above)
Comment (PetrovKP) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp]: It's a pity, it would be nice.

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.h [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+            _finalBetas->getBlockOfRows(0, _nBetas, daal_dm::writeOnly, bFinalBetas);
+
+            auto src = bBetas.getBlockPtr();
+            auto dst = bFinalBetas.getBlockPtr();
+
+            for (size_t i = 0; i < _nBetas; i++)
+            {
+                for (size_t j = 0; j < _nFeatures; j++)
+                {
+                    dst[i*_nFeatures + j] = src[i*_nBetaCols + j + 1];
```
Comment (PetrovKP) [inline examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.h:176]: [minor] For good, this code should be on sycl.

File: examples/daal/cpp_sycl/source/logistic_regression/log_reg_dense_online.cpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    /* Create an algorithm object to predict values of logistic regression */
+    LogRegPrediction algorithm(nClasses);
+    algorithm.parameter().resultsToEvaluate |= classifier::computeClassProbabilities | classifier::computeClassLogProbabilities;
+
+    /* Pass a testing data set and the trained model to the algorithm */
+    algorithm.input.set(class_prediction::data, testX);
+    algorithm.input.set(class_prediction::model, model);
+
+    /* Predict values of logistic regression */
+    algorithm.compute();
```
Comment (PetrovKP) [inline examples/daal/cpp_sycl/source/logistic_regression/log_reg_dense_online.cpp:133]: [minor] prediction would also be good to show online, especially it's very easy to follow.

File: examples/daal/cpp_sycl/source/logistic_regression/log_reg_dense_online.cpp [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+        testModel(model);
+    }
+
+    return 0;
+}
+
+logistic_regression::ModelPtr trainModel()
+{
+    constexpr bool isIntercept           = false;
```
Comment (PetrovKP) [inline examples/daal/cpp_sycl/source/logistic_regression/log_reg_dense_online.cpp:75]: [style] Cool of course, but in other examples, no one does that. Looks probably not consistency.

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.h [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+            for (size_t i = 0; i < bd.getNumberOfRows(); i += nBetaCols)
+            {
+                ptr[i] = Fptype(1e-3);
+            }
+
+            _workPoint->releaseBlockOfRows(bd);
+        }
+    }
+
+    void checkParameterDomains(size_t nClasses,
```
Comment (PetrovKP) [inline examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.h:225]: [minor] I think this is superfluous - we are not writing a full-product class/algorithm, but just an simple example, otherwise the checks will not be enough

**Review by PetrovKP — APPROVED** [review]

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp (hunk 2 above)
Comment (michael-smirnov) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.hpp]: It's done with intent to have default ctor of the class - this avoids user the need to create multiple objects to solve the tasks with different parameters, and it's easy to initialize user-owned classes which hold `LogisticRegressionOnline` object - no need to store it by pointer if we have a default ctor.

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.h (hunk 8 above)
Comment (michael-smirnov) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.h:176]: It's done this way for simplicity - if you insist on the change, I can apply it

File: examples/daal/cpp_sycl/source/logistic_regression/log_reg_dense_online.cpp (hunk 9 above)
Comment (michael-smirnov) [reply examples/daal/cpp_sycl/source/logistic_regression/log_reg_dense_online.cpp:133]: It's done this way for simplicity - if you insist on the change, I can apply it

File: examples/daal/cpp_sycl/source/logistic_regression/log_reg_dense_online.cpp (hunk 10 above)
Comment (michael-smirnov) [reply examples/daal/cpp_sycl/source/logistic_regression/log_reg_dense_online.cpp:75]: Where do you see the inconsistency?

File: examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.h (hunk 11 above)
Comment (michael-smirnov) [reply examples/daal/cpp_sycl/source/logistic_regression/logistic_regression_online.h:225]: Why do you think these checks are not enough?

**Review by KulikovNikita — APPROVED** [review]
